### PR TITLE
Add the ability to run dgoss when the docker daemon is not local. 

### PR DESCRIPTION
**Pull request by plispe**
I created pull request which is discussed in #269 and #243. All the changes should be backwards compatible.

### CODE REVIEW

**Comment by aelsabbahy** [comment]
I'm guessing the #243 link was a mistake.

Skimmed through the code and I like this approach. I'll test it and provide feedback this week.

File: extras/dgoss/dgoss [hunk 1]
```diff
@@ -25,9 +25,26 @@ run(){
...
+        ;;
+      cp)
+        info "Creating docker container"
+        id=$(docker create "${@:2}")
+        info "Copy goss files into container"
+        docker cp $tmp_dir $id:/goss
+        info "Starting docker container"
+        docker start $id > /dev/null
+        ;;
+      *) error "Wrong goss files strategy used! Correct options are \"mount\" of \"cp\"."
```
Comment (elliotweiser) [inline extras/dgoss/dgoss]: mount _or_ copy

**Comment by elliotweiser** [comment]
It should be possible to copy the `goss.yaml` file out of the running docker container when using the `copy` option, no?

**Comment by plispe** [comment]
@elliotweiser I think, that goss.yaml is copied only when you use `dgoss edit` not when `dgoss run`  is used. The  `dgoss edit` should work as expected.

File: extras/dgoss/dgoss (hunk 1 above)
Comment (aelsabbahy) [inline extras/dgoss/dgoss]: Cp is fine, I think it's the "of" instead of "or" that needs fixing.

Comment (plispe) [inline extras/dgoss/dgoss]: yes. "of' is a typo. sorry for that.

**Comment by elliotweiser** [comment]
You are correct about the `goss.yaml` not being copied out of the container when using `dgoss run`. 

I guess I'm responding to this statement:

> With the `'cp'` strategy you lose the ability to write tests or waits against the docker output.

Since the generated `goss.yaml` would normally live on the volume, retrieving it after `dgoss edit` is trivial. When using the `cp` mode, I think the original behavior can be preserved by executing a `docker cp <CONTAINER>:file file`. 

(I personally can't think of a good reason for developing tests against a remote docker daemon, but for the sake of thorough code review, I thought I'd point this out).

Overall, LGTM 👍

File: extras/dgoss/dgoss (hunk 1 above)
Comment (elliotweiser) [inline extras/dgoss/dgoss]: Yep just a typo: I meant `of` vs `or`.

Whatever your preference is for `copy` vs `cp` ;)

Comment (elliotweiser) [inline extras/dgoss/dgoss]: Sorry about that... my brain did an auto-complete 😅

**Comment by plispe** [comment]
> Since the generated goss.yaml would normally live on the volume, retrieving it after dgoss edit is trivial. When using the cp mode, I think the original behaviour can be preserved by executing a docker cp <CONTAINER>:file file.

There is actually no volume used with `dgoss edit` but, the generated `goss.yaml` is copied with `docker cp` command from the container.  Therefore `dgoss edit` is unaffected by my changes.

> With the 'cp' strategy you lose the ability to write tests or waits against the docker output.

This statement is about the following command  from `dgoss` script `docker logs -f "$id" > "$tmp_dir/docker_output.log" 2>&1 &` whose behaviour can not be preserved with `cp` strategy. Because you have not possibility to stream `docker log -f` output to the running container without a mounted volume.

**Comment by elliotweiser** [comment]
I stand corrected: `dgoss edit` indeed doesn't use docker volumes. In which case...

:shipit: !

**Comment by plispe** [comment]
@elliotweiser 

> (I personally can't think of a good reason for developing tests against a remote docker daemon, but for the sake of thorough code review, I thought I'd point this out).

The only use-case could be the use of [CicleCI Debugging Jobs Over SSH] (https://circleci.com/docs/2.0/ssh-access-jobs/#nav-button) as CircleCI runs jobs in [separated environments](https://circleci.com/docs/2.0/building-docker-images/#separation-of-environments). I debug sometimes my build over ssh.

**Comment by elliotweiser** [comment]
Neat! I'll have to check that out. Thanks for the contribution and the thorough responses!

**Comment by plispe** [comment]
you are welcome

**Comment by aelsabbahy** [comment]
Apologies for the delay on this, looks great. Thanks for the contribution!

**Comment by plispe** [comment]
No problem. I used my fork untill now. Thanks for merge.

**Pull request merged**